src/pdal/PyArray.cpp

```diff
@@ -164,51 +164,85 @@ Array::~Array()
     Py_XDECREF(m_array);
 }
 
-
-ArrayIter& Array::iterator()
+std::shared_ptr<ArrayIter> Array::iterator()
 {
-    ArrayIter *it = new ArrayIter(m_array);
-    m_iterators.emplace_back((it));
-    return *it;
+    return std::make_shared<ArrayIter>(m_array, m_stream_handler);
 }
 
+ArrayIter::ArrayIter(PyArrayObject* np_array, std::shared_ptr<ArrayStreamHandler> stream_handler)
+    : m_stream_handler(std::move(stream_handler))
+{
+    resetIterator(np_array);
+}
 
-ArrayIter::ArrayIter(PyArrayObject* np_array)
+void ArrayIter::resetIterator(std::optional<PyArrayObject*> np_array = {})
 {
-    m_iter = NpyIter_New(np_array,
-        NPY_ITER_EXTERNAL_LOOP | NPY_ITER_READONLY | NPY_ITER_REFS_OK,
-        NPY_KEEPORDER, NPY_NO_CASTING, NULL);
-    if (!m_iter)
-        throw pdal_error("Unable to create numpy iterator.");
+    std::optional<int> stream_chunk_size = std::nullopt;
+    if (m_stream_handler) {
+        stream_chunk_size = (*m_stream_handler)();
```

Comment:
I've changed type to `int64_t` instead of `uint64_t` just for the sake of presenting a clearer error message to the python developer in case they mistakenly return a negative number.

Will be this:
`RuntimeError: Stream chunk size not in the range of array length: -1`

Instead of this:
`RuntimeError: Unable to cast Python instance of type <class 'int'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)`